### PR TITLE
MIDI-64: Loss-reduction

### PR DESCRIPTION
**Pull request by SamuelJanas**
[MIDI-64](https://sacrebleu.atlassian.net/browse/MIDI-64)

[MIDI-64]: https://sacrebleu.atlassian.net/browse/MIDI-64?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by SamuelJanas** [comment]
Started logging gradients to monitor the difference between reduction="sum" and reduction="mean", 
For mean - [wandb](https://wandb.ai/sjanas/autoencoders/runs/qpqs13ku?workspace=user-sneakyinsect)
For sum -  [wandb](https://wandb.ai/sjanas/autoencoders/runs/rfmsoc32?workspace=user-sneakyinsect)
Looks like the gradient is vanishing

**Comment by roszcz** [comment]
That's intriguing - do you know if the change is doing anything other than dividing the final value by a constant? 🤔

**Comment by SamuelJanas** [comment]
> That's intriguing - do you know if the change is doing anything other than dividing the final value by a constant? 🤔

According to [this discussion](https://discuss.pytorch.org/t/bce-loss-reduction/31375/2) on pytorch forum it doesn't do anything else.
Tested it with
```Python
    recon_batch, mu, logvar = model(data)
    recon_loss = F.binary_cross_entropy(recon_batch, data, reduction="mean")
    recon_loss = recon_loss * data.shape[1] * data.shape[2] * data.shape[3] * 32 # magic number here is batch size
    KLD = -0.5 * torch.mean(1 + logvar - mu.pow(2) - logvar.exp())

    loss = recon_loss + KLD
```
and the [results](https://wandb.ai/sjanas/autoencoders/runs/osomfnqm?workspace=user-sneakyinsect) are the same as `reduction="sum"` so empirically there is no difference either.

**Comment by SamuelJanas** [comment]
Something I noticed, when you look at [wandb-mean](https://wandb.ai/sjanas/autoencoders/runs/qpqs13ku?workspace=user-sneakyinsect). It seems like only the encoder gradients are zeroed. That still doesn't help much given that it uses LeakyReLU **and** batch norm, which should make it more resistant to vanishing gradient. 

At first I thought that maybe the sum reduction is just inherent with VAEs but looking at [this](https://github.com/AntixK/PyTorch-VAE/blob/master/models/vanilla_vae.py) implementation it clearly isn't the case. There are some papers that use scalars on recon and KLD losses but I don't think this is the issue here either

**Comment by SamuelJanas** [comment]
Some things I tried to mitigate the vanishing:
- kaiming init
- tweak learning rate
- use MSE instead of BCE

**Comment by roszcz** [comment]
One thing to try is to compare the gradients with the regular autoencoder, maybe this will give us some hints. I think it's a good exercise to investigate further, but let me know if you'll feel stuck :)

**Comment by SamuelJanas** [comment]
> One thing to try is to compare the gradients with the regular autoencoder, maybe this will give us some hints. I think it's a good exercise to investigate further, but let me know if you'll feel stuck :)

Wanted to do just that, trained an AE to have the weight but I found out that it doesn't work as intended:
image
Did you have a similar issue after rewriting the code?

**Comment by roszcz** [comment]
There was a scaling bug in the AE evaluation code, after fixing it this is what I get:

image

**Comment by SamuelJanas** [comment]
Tried visualizing despite the artifact AE produces, which probably doesn't influence weights by much, it's still better than VAE with mean reduction. Here are some results:
## VAE vs AE (mean reduction)
Hard to show the scale this way, will be visible better in another part
encoder gradients

decoder gradients

## VAE vs AE (sum reduction)
Model trained with sum reduction but scaled afterwards for visualization, scale factor is batch_size*28*28
encoder gradients scaled

decoder gradients scaled


The weights are similar in a way, the gradient vanish persists unfortunately. My suggestion is that we should keep the sum reduction for the loss function but log it's scaled value, this way the model will train as it should and we get all the benefits of not looking at huge numbers

**Comment by roszcz** [comment]
Thanks, I think we can leave it at that 👍 

Can you rebase this branch and get rid of the conflicts? :)

**Comment by SamuelJanas** [comment]
I believe we can close the PR now

**Comment by roszcz** [comment]
One thing that I'm still wondering about, is that we have sum reduction for reconstruction, but mean for KLD:

```python
    recon_loss = F.binary_cross_entropy(recon_batch, data, reduction="sum")
    KLD = -0.5 * torch.mean(1 + logvar - mu.pow(2) - logvar.exp())
```

🤔

**Comment by SamuelJanas** [comment]
> One thing that I'm still wondering about, is that we have sum reduction for reconstruction, but mean for KLD:
> 
> ```python
>     recon_loss = F.binary_cross_entropy(recon_batch, data, reduction="sum")
>     KLD = -0.5 * torch.mean(1 + logvar - mu.pow(2) - logvar.exp())
> ```
> 
> 🤔

Since loss is just the combination of both
```Python
    loss = recon_loss + KLD
```
the mean there is there for the scale, KLD measures how much learned distribution deviates from the prior. I don't think using sum there would make much sense

**Review by roszcz — APPROVED** [review]

**Comment by SamuelJanas** [comment]
In a form of experiment:
```Python
    recon_loss = F.binary_cross_entropy(recon_batch, data, reduction="mean")
    KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
```
gradient vanished,
```Python
    recon_loss = F.binary_cross_entropy(recon_batch, data, reduction="sum")
    KLD = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
```
The results are slightly worse, 
image
KLD is steadily growing, I'll check that in a second on a longer training, this might motivate the VAE to focus on distribution more and less on decoders just being good.

**Comment by SamuelJanas** [comment]
> I'll check that in a second on a longer training, this might motivate the VAE to focus on distribution more and less on decoders just being good.

This simply made trainig more unstable, mean makes the KLD smoother, which as a result allows for faster convergance and more stable training. Looking at the results also confirms that using mean just works better 😅 

## sum
image

## mean with half of the training done:
image

**Comment by SamuelJanas** [comment]
Well... I tried but I don't think that rebasing should look the way I did it. 😖 Sorry for possible trouble but @roszcz could you please take a loot at this and well possibly tell me what I'm doing wrong? 
Following your suggestion
> Thanks, I think we can leave it at that 👍
>Can you rebase this branch and get rid of the conflicts? :)

I wanted to `git rebase master` to rebase MIDI-64 onto master but now I don't think that's correct?
Thanks in advance for your patiance that would be my first rebase ever I believe 😅

**Comment by roszcz** [comment]
@SamuelJanas When looking at the "Files changed" tab in this PR, if these are all the changes you actually did on this branch, that's great. Let me know if that's the case, if not, we'll look for any lost changes 🕵️

**Comment by SamuelJanas** [comment]
> @SamuelJanas When looking at the "Files changed" tab in this PR, if these are all the changes you actually did on this branch, that's great. Let me know if that's the case, if not, we'll look for any lost changes 🕵️

I believe it should be all of them, there weren't any important changes. We settled on using the code as is and added gradient logging just to be sure it doesn't vanish
